book/src/leader-validator-transition.md

```diff
@@ -0,0 +1,129 @@
+
+1. Keep track of *time* as defined by PoH height and stop the validator when it
+hits the scheduled leader slot.
+
+2. Reset the clock whenever a validator votes.  The last validator vote is the
+validators most recent commitment. It is therefore the closest fork to the
+scheduled slot.
+
+3. Provide the entries that connect the last vote to the scheduled slot.
+
+The interface to this object implements the following methods:
+
+* new - Start the PoH thread, and compute it until end\_ticks.  This method
+would take the starting hash the starting tick height and ending tick height and
+create a PoH thread that runs until the end.
+
+* restart - Restart the PoH thread at a new hash and reset height.  It still
+runs until the configured ending tick height.
+
+* final\_entries - Get the entries that connect the set of entries from the
+restart point to the final tick as specified by new.
+
+### PoH Recorder
+
+This object handles PoH recording for a Leader.  It solves the following
+problems:
+
+1. Keep track of *time* as defined by PoH height, and stop the leader when it
+gets to the end of its scheduled leader slot.
+
+2. Register ticks with the BankFork as they are produced by PoH.
+
+3. Record entries that are produced by the TPU into PoH.
+
+* new - Create a new recorder, with the starting entries as generated by the
+Generator.  The recorder runs until the ending tick height, and registers ticks
+with a BankFork.
+
+* record - Adds a vector of transactions to PoH, and register the tick with BankFork.
+
+## Fullnode Loop
+
+This object manages the transition between modes.  The main idea is that once a
+ledger is replayed, the validator can run until there is proof that it can be a
+leader.  The Leader can then run and record its transactions.
+
+The core idea is that the loop is synchronized to PoH, and does a synchronous
+start and stop of a validator and leader while having a reference to all the
+bank forks and the leader\_scheduler.  The following is pseudocode for the
+loop:
+
+The fullnode loop:
+
+These components are passed in into the loop from the outside.
+
+* forks - The Forks generated by replaying the ledger.
+
+* leader\_scheduler - The leader\_scheduler generated by replaying the ledger.
+
+* cluster\_info -  The network. The code to continuously receive and retransmit
+blocks is owned by this.
+
+* block\_tree -  The ledger.
+
+loop:
+
+1. Get my next scheduled slot.
+
+2. Start the PoH Generator to run until the next scheduled slot.
+
+3. Run the TVU over all the forks.
+
+   1. TVU will send votes for votable forks.
+
+4. Wait for PoH generator to finish.
+
+5. Stop the TVU.
+
+6. Get the entries connecting a voted slot in Forks to the scheduled leader slot.
+
+7. Start a PoH recorder for the leader slot range of ticks.
+
+8. Start the TPU over a BankFork checkpoint for the leader slot.
+
+9. Wait for the PoH recorder to finish.
+
+   1. TPU will freeze BankFork.  The vote will be sent in 3.
+
+10. Goto 1.
+
+## BankFork
```

Comment:
Note to self, move this up to someplace before it is first referenced in this document.